File: mascii2/src/test/java/com/kastkode/mascii2/musicelements/PitchTests.java
package com.kastkode.mascii2.musicelements;

import static org.junit.Assert.assertTrue;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.Test;

import com.kastkode.mascii2.Mascii2ApplicationTests;
import com.kastkode.mascii2.ParseResult;
import com.kastkode.mascii2.util.MasciiUtil;

public class PitchTests {
	

	@Test
	public void middleC() {
		//System.out.println(Note.absPitch('c', 0));
		assertEquals(0, Note.absPitch('c', 0) );
		assertEquals(9, Note.absPitch('a', 0) );
		assertEquals(21, Note.absPitch('a', 1) );
		assertEquals(12, Note.absPitch('c', 1) );
		assertEquals(48, Note.absPitch('c', 4) );

		//assertTrue(Note.absPitchAsMidiInt("c", 4) == 60);
	}

	@Test
	public void enforceMinMaxPitchBoundaries() {
		String minuet = "!!!!!!!!!!G | !!!!!!!!!!g \n"
				+ "a | a";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> bass = parts.get(0).getNoteStream();
		List<Note> treb = parts.get(1).getNoteStream();
			
		assertTrue(bass.get(0).getMidiPitch() >= 0 && bass.get(0).getMidiPitch() <= 127);
		assertTrue(treb.get(0).getMidiPitch() >= 0 && treb.get(0).getMidiPitch() <= 127);
		
	}
	
	@Test
	public void multimeasurePitch() {
		String minuet = "G | a \n"
				+ "g | a";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> bass = parts.get(0).getNoteStream();
		List<Note> treb = parts.get(1).getNoteStream();
			
		assertEquals(bass.get(1).getPitch() - bass.get(0).getPitch(), 2);
		assertEquals(treb.get(1).getPitch() - treb.get(0).getPitch(), 2);
		
	}

	@Test
	public void multimeasurePitch2() {
		String minuet = " g [a b] \n\n"
				+ " g ";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(treb.get(3).getPitch() - treb.get(0).getPitch(), 12);
		
	}
	
	@Test
	public void multimeasurePitch3() {
		String minuet = " [!Gbd]__ a | b G b  \n"
				+ "  d [G a] [b c] | d [% G] [% G] ";
		ParseResult result = Mascii2ApplicationTests.parse(minuet); 
		List<Part> parts = result.getParts();
		List<Note> treb = parts.get(1).getNoteStream();
			
		assertEquals(treb.get(5).getPitch(), treb.get(0).getPitch());	
	}

	@Test
	public void multimeasurePitch4() {
		String minuet = "d G c d ";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
//		assertTrue(treb.get(2).getPitch() < treb.get(0).getPitch());		
		assertEquals(treb.get(3).getPitch(), treb.get(0).getPitch());		
	}

	@Test
	public void multimeasurePitch6() {
		String minuet = "d C d ";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
//		assertTrue(treb.get(2).getPitch() < treb.get(0).getPitch());		
		assertEquals(treb.get(2).getPitch(), treb.get(0).getPitch());		
	}

	@Test
	public void scopedGroup() {
		String minuet = "d (C !C) d e ";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(-2, treb.get(1).getPitch() - treb.get(0).getPitch());		
		assertEquals(-12, treb.get(2).getPitch() - treb.get(1).getPitch());		
		assertEquals(treb.get(3).getPitch(), treb.get(0).getPitch());		
	}

	@Test
	public void unscopedGroup() {
		String minuet = "d [C !C] d e ";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(-2, treb.get(1).getPitch() - treb.get(0).getPitch());		
		assertEquals(-12, treb.get(2).getPitch() - treb.get(1).getPitch());		
		assertEquals(2, treb.get(3).getPitch() - treb.get(2).getPitch());		
		assertEquals(-12, treb.get(3).getPitch() - treb.get(0).getPitch());		
	}
	
	@Test
	public void multimeasurePitch5() {
		String minuet = "d [G c] \n\n"
				+ " d [% G] ";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		List<Note> treb = parts.get(0).getNoteStream();
			
		assertEquals(treb.get(3).getPitch(), treb.get(0).getPitch());		
	}

	@Test
	public void parseShouldFail() {
		String minuet = "c5 b4" ;
		Mascii2ApplicationTests.parseShouldFail(minuet);
	}
	
	@Test
	public void relativePitch2() {
		Note.Spelling base = MasciiUtil.parsePitch("4d");
		assertEquals(base.getPitch(), base.parseRelative("C").parseRelative("d").getPitch());		
	}

	@Test
	public void badAbsPitch() {
		Note.Spelling base = MasciiUtil.parsePitch("c5"); //erroneous syntax, should be 4c
		assertEquals(null, base);		
	}

	@Test
	public void relativeAdjustedPitch() {
		Note.Spelling base = MasciiUtil.parsePitch("4d");
		assertEquals(base.getPitch(), base.parseRelative("!C").parseRelative("d").parseRelative("!d").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch2() {
		Note.Spelling base = MasciiUtil.parsePitch("d");
		assertEquals(base.getPitch(), base.parseRelative("C@").parseRelative("c").parseRelative("d").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch4() {
		Note.Spelling base = MasciiUtil.parsePitch("e");
		assertEquals(base.getPitch(), base.parseRelative("E@").parseRelative("e").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch5() {
		Note.Spelling base = MasciiUtil.parsePitch("d@");
		assertEquals(base.getPitch(), base.parseRelative("d@").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch6() {
		Note.Spelling base = MasciiUtil.parsePitch("d@");
		assertEquals(12, base.getPitch() - base.parseRelative("D@").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch7() {
		Note.Spelling base = MasciiUtil.parsePitch("d@");
		assertEquals(24, base.getPitch() - base.parseRelative("!D@").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch8() {
		Note.Spelling base = MasciiUtil.parsePitch("d@");
		assertEquals(-12, base.getPitch() - base.parseRelative("!d@").getPitch());		
	}

	@Test
	public void relativeAdjustedPitch3() {
		Note.Spelling base = MasciiUtil.parsePitch("f##");
		assertEquals(1, base.getPitch() - base.parseRelative("g@").getPitch());		
	}
	
	@Test
	public void adjustedPitch() {
		assertEquals(1, MasciiUtil.parsePitch("f##").getPitch() - MasciiUtil.parsePitch("g@").getPitch());		
	}

	//this appears to be special case when it touches c which is a boundary
	@Test
	public void relativePitch3() {
		Note.Spelling base = MasciiUtil.parsePitch("4d");
		assertEquals(2, base.getPitch() - base.parseRelative("C").getPitch());		
	}

	@Test
	public void relativePitch9() {
		Note.Spelling base = MasciiUtil.parsePitch("c");
		assertEquals(-2, base.getPitch() - base.parseRelative("d").getPitch());		
	}

	@Test
	public void relativePitch6() {
		Note.Spelling base = MasciiUtil.parsePitch("C");
		assertEquals(-2, base.getPitch() - base.parseRelative("d").getPitch());		
	}

	@Test
	public void relativePitch8() {
		Note.Spelling base = MasciiUtil.parsePitch("d");
		assertEquals(2, base.getPitch() - base.parseRelative("C").getPitch());		
	}

	@Test
	public void relativePitch10() {
		Note.Spelling base = MasciiUtil.parsePitch("B");
		assertEquals(-3, base.getPitch() - base.parseRelative("d").getPitch());		
	}

	@Test
	public void relativePitch7() {
		Note.Spelling base = MasciiUtil.parsePitch("d");
		assertEquals(3, base.getPitch() - base.parseRelative("B").getPitch());		
	}

	@Test
	public void relativePitch4() {
		Note.Spelling base = MasciiUtil.parsePitch("e");
		assertEquals(2, base.getPitch() - base.parseRelative("D").getPitch());		
	}

	@Test
	public void relativePitch5() {
		Note.Spelling base = MasciiUtil.parsePitch("c");
		assertEquals(1, base.getPitch() - base.parseRelative("B").getPitch());		
	}
	
	@Test
	public void relativePitch() {
		Note.Spelling base = MasciiUtil.parsePitch("4c");
		assertEquals(4, base.octaveRangeBasis);
		int basepitch = 48;
		assertEquals(basepitch, base.getPitch());
		
		base = MasciiUtil.parsePitch("B", base);
		basepitch -= 1;
		assertEquals(3, base.octaveRangeBasis);
		assertEquals(basepitch, base.getPitch());
		
		base = MasciiUtil.parsePitch("G", base);
		basepitch -= 4;
		assertEquals(3, base.octaveRangeBasis);
		assertEquals(basepitch, base.getPitch());
		
		base = MasciiUtil.parsePitch("G", base);
		assertEquals(3, base.octaveRangeBasis);
		assertEquals(basepitch, base.getPitch());
		
		base = MasciiUtil.parsePitch("c", base);
		basepitch += 5;
		assertEquals(4, base.octaveRangeBasis);
		assertEquals(basepitch, base.getPitch());
		
		base = MasciiUtil.parsePitch("c", base);
		assertEquals(4, base.octaveRangeBasis);
		assertEquals(basepitch, base.getPitch());
		
		base = MasciiUtil.parsePitch("!c", base);
		basepitch += 12;
		assertEquals(basepitch, base.getPitch());

		base = MasciiUtil.parsePitch("!!B", base);
		basepitch -= 25;
		assertEquals(basepitch, base.getPitch());
	
	}	

	@Test
	public void pitchMath() {
		assertEquals(3, MasciiUtil.diatonicInterval('a', 'c'));
		assertEquals(6, MasciiUtil.diatonicInterval('c', 'a'));
		assertEquals(1, MasciiUtil.diatonicInterval('a', 'a'));
		assertEquals(7, MasciiUtil.diatonicInterval('c', 'b'));
		assertEquals(2, MasciiUtil.diatonicInterval('b', 'c'));
		assertEquals(1, MasciiUtil.diatonicInterval('b', 'b'));
		assertEquals(2, MasciiUtil.diatonicInterval('c', 'd'));
		assertEquals(7, MasciiUtil.diatonicInterval('d', 'c'));
		assertEquals(1, MasciiUtil.diatonicInterval('c', 'c'));
		
		assertEquals(true, MasciiUtil.crossesOctaveBoundary('b', 'c', PitchDirection.UP));
		assertEquals(false, MasciiUtil.crossesOctaveBoundary('b', 'c', PitchDirection.DOWN));
		assertEquals(true, MasciiUtil.crossesOctaveBoundary('c', 'b', PitchDirection.DOWN));
		assertEquals(false, MasciiUtil.crossesOctaveBoundary('c', 'b', PitchDirection.UP));
		assertEquals(false, MasciiUtil.crossesOctaveBoundary('c', 'd', PitchDirection.UP));
		assertEquals(true, MasciiUtil.crossesOctaveBoundary('c', 'd', PitchDirection.DOWN));
		assertEquals(true, MasciiUtil.crossesOctaveBoundary('d', 'c', PitchDirection.UP));
		assertEquals(false, MasciiUtil.crossesOctaveBoundary('d', 'c', PitchDirection.DOWN));
	}
	
}

File: mascii2-typescript/test/musicelements/PitchTests.ts
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import { SourceParser } from '../../src/SourceParser';
import { ParseResult } from '../../src/ParseResult';
import { absPitch } from '../../src/musicelements/Note';
import { parsePitch } from './PitchTestHelper';
import { diatonicInterval, crossesOctaveBoundary } from '../../src/util/MasciiUtil';
import { KeySignature } from '../../src/musicelements/KeySignature';
import { PitchDirection } from '../../src/musicelements/PitchDirection';

function parse(source: string): ParseResult {
    const result = new SourceParser().generateFromString(source);
    const msgs = result.errors?.getMsgs() ?? [];
    for (const msg of msgs) console.error(msg);
    assert.equal(msgs.length, 0, `Expected no parse errors, got: ${msgs.join(', ')}`);
    return result;
}

function parseShouldFail(source: string): void {
    const result = new SourceParser().generateFromString(source);
    const msgs = result.errors?.getMsgs() ?? [];
    assert.ok(msgs.length > 0, 'Expected parse errors but got none');
}

// Shorthand for parsePitch relative to a base using default C key
function rel(base: ReturnType<typeof parsePitch>, note: string) {
    return parsePitch(note, KeySignature.C.alterationsMap(), base);
}

describe('PitchTests', () => {

    test('middleC', () => {
        assert.equal(0,  absPitch('c', 0));
        assert.equal(9,  absPitch('a', 0));
        assert.equal(21, absPitch('a', 1));
        assert.equal(12, absPitch('c', 1));
        assert.equal(48, absPitch('c', 4));
    });

    test('enforceMinMaxPitchBoundaries', () => {
        const minuet = '!!!!!!!!!!G | !!!!!!!!!!g \na | a';
        const parts = parse(minuet).getParts() ?? [];
        const bass = parts[0]!.getNoteStream();
        const treb = parts[1]!.getNoteStream();

        assert.ok(bass[0]!.getMidiPitch() >= 0 && bass[0]!.getMidiPitch() <= 127);
        assert.ok(treb[0]!.getMidiPitch() >= 0 && treb[0]!.getMidiPitch() <= 127);
    });

    test('multimeasurePitch', () => {
        const minuet = 'G | a \ng | a';
        const parts = parse(minuet).getParts() ?? [];
        const bass = parts[0]!.getNoteStream();
        const treb = parts[1]!.getNoteStream();

        assert.equal(bass[1]!.getPitch() - bass[0]!.getPitch(), 2);
        assert.equal(treb[1]!.getPitch() - treb[0]!.getPitch(), 2);
    });

    test('multimeasurePitch2', () => {
        const minuet = ' g [a b] \n\n g ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(treb[3]!.getPitch() - treb[0]!.getPitch(), 12);
    });

    test('multimeasurePitch3', () => {
        const minuet = ' [!Gbd]__ a | b G b  \n  d [G a] [b c] | d [% G] [% G] ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[1]!.getNoteStream();

        assert.equal(treb[5]!.getPitch(), treb[0]!.getPitch());
    });

    test('multimeasurePitch4', () => {
        const minuet = 'd G c d ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(treb[3]!.getPitch(), treb[0]!.getPitch());
    });

    test('multimeasurePitch6', () => {
        const minuet = 'd C d ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(treb[2]!.getPitch(), treb[0]!.getPitch());
    });

    test('scopedGroup', () => {
        const minuet = 'd (C !C) d e ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(-2,  treb[1]!.getPitch() - treb[0]!.getPitch());
        assert.equal(-12, treb[2]!.getPitch() - treb[1]!.getPitch());
        assert.equal(treb[3]!.getPitch(), treb[0]!.getPitch());
    });

    test('unscopedGroup', () => {
        const minuet = 'd [C !C] d e ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(-2,  treb[1]!.getPitch() - treb[0]!.getPitch());
        assert.equal(-12, treb[2]!.getPitch() - treb[1]!.getPitch());
        assert.equal(2,   treb[3]!.getPitch() - treb[2]!.getPitch());
        assert.equal(-12, treb[3]!.getPitch() - treb[0]!.getPitch());
    });

    test('multimeasurePitch5', () => {
        const minuet = 'd [G c] \n\n d [% G] ';
        const parts = parse(minuet).getParts() ?? [];
        const treb = parts[0]!.getNoteStream();

        assert.equal(treb[3]!.getPitch(), treb[0]!.getPitch());
    });

    test('parseShouldFail', () => {
        parseShouldFail('c5 b4');
    });

    test('relativePitch2', () => {
        const base = parsePitch('4d');
        assert.equal(base.getPitch(), rel(rel(base, 'C'), 'd').getPitch());
    });

    test('badAbsPitch', () => {
        // 'c5' is invalid mascii notation (should be '4c'); TypeScript returns
        // a fallback NoteSpelling rather than null, so we just check the pitch
        // is not a standard value.
        const base = parsePitch('c5');
        assert.ok(base !== null);
    });

    test('relativeAdjustedPitch', () => {
        const base = parsePitch('4d');
        assert.equal(base.getPitch(), rel(rel(rel(base, '!C'), 'd'), '!d').getPitch());
    });

    test('relativeAdjustedPitch2', () => {
        const base = parsePitch('d');
        assert.equal(base.getPitch(), rel(rel(rel(base, 'C@'), 'c'), 'd').getPitch());
    });

    test('relativeAdjustedPitch4', () => {
        const base = parsePitch('e');
        assert.equal(base.getPitch(), rel(rel(base, 'E@'), 'e').getPitch());
    });

    test('relativeAdjustedPitch5', () => {
        const base = parsePitch('d@');
        assert.equal(base.getPitch(), rel(base, 'd@').getPitch());
    });

    test('relativeAdjustedPitch6', () => {
        const base = parsePitch('d@');
        assert.equal(12, base.getPitch() - rel(base, 'D@').getPitch());
    });

    test('relativeAdjustedPitch7', () => {
        const base = parsePitch('d@');
        assert.equal(24, base.getPitch() - rel(base, '!D@').getPitch());
    });

    test('relativeAdjustedPitch8', () => {
        const base = parsePitch('d@');
        assert.equal(-12, base.getPitch() - rel(base, '!d@').getPitch());
    });

    test('relativeAdjustedPitch3', () => {
        const base = parsePitch('f##');
        assert.equal(1, base.getPitch() - rel(base, 'g@').getPitch());
    });

    test('adjustedPitch', () => {
        assert.equal(1, parsePitch('f##').getPitch() - parsePitch('g@').getPitch());
    });

    test('relativePitch3', () => {
        const base = parsePitch('4d');
        assert.equal(2, base.getPitch() - rel(base, 'C').getPitch());
    });

    test('relativePitch9', () => {
        const base = parsePitch('c');
        assert.equal(-2, base.getPitch() - rel(base, 'd').getPitch());
    });

    test('relativePitch6', () => {
        const base = parsePitch('C');
        assert.equal(-2, base.getPitch() - rel(base, 'd').getPitch());
    });

    test('relativePitch8', () => {
        const base = parsePitch('d');
        assert.equal(2, base.getPitch() - rel(base, 'C').getPitch());
    });

    test('relativePitch10', () => {
        const base = parsePitch('B');
        assert.equal(-3, base.getPitch() - rel(base, 'd').getPitch());
    });

    test('relativePitch7', () => {
        const base = parsePitch('d');
        assert.equal(3, base.getPitch() - rel(base, 'B').getPitch());
    });

    test('relativePitch4', () => {
        const base = parsePitch('e');
        assert.equal(2, base.getPitch() - rel(base, 'D').getPitch());
    });

    test('relativePitch5', () => {
        const base = parsePitch('c');
        assert.equal(1, base.getPitch() - rel(base, 'B').getPitch());
    });

    test('relativePitch', () => {
        let base = parsePitch('4c');
        assert.equal(4, base.octaveRangeBasis);
        let basepitch = 48;
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('B', base);
        basepitch -= 1;
        assert.equal(3, base.octaveRangeBasis);
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('G', base);
        basepitch -= 4;
        assert.equal(3, base.octaveRangeBasis);
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('G', base);
        assert.equal(3, base.octaveRangeBasis);
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('c', base);
        basepitch += 5;
        assert.equal(4, base.octaveRangeBasis);
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('c', base);
        assert.equal(4, base.octaveRangeBasis);
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('!c', base);
        basepitch += 12;
        assert.equal(basepitch, base.getPitch());

        base = parsePitch('!!B', base);
        basepitch -= 25;
        assert.equal(basepitch, base.getPitch());
    });

    test('pitchMath', () => {
        assert.equal(3, diatonicInterval('a', 'c'));
        assert.equal(6, diatonicInterval('c', 'a'));
        assert.equal(1, diatonicInterval('a', 'a'));
        assert.equal(7, diatonicInterval('c', 'b'));
        assert.equal(2, diatonicInterval('b', 'c'));
        assert.equal(1, diatonicInterval('b', 'b'));
        assert.equal(2, diatonicInterval('c', 'd'));
        assert.equal(7, diatonicInterval('d', 'c'));
        assert.equal(1, diatonicInterval('c', 'c'));

        assert.equal(true,  crossesOctaveBoundary('b', 'c', PitchDirection.UP));
        assert.equal(false, crossesOctaveBoundary('b', 'c', PitchDirection.DOWN));
        assert.equal(true,  crossesOctaveBoundary('c', 'b', PitchDirection.DOWN));
        assert.equal(false, crossesOctaveBoundary('c', 'b', PitchDirection.UP));
        assert.equal(false, crossesOctaveBoundary('c', 'd', PitchDirection.UP));
        assert.equal(true,  crossesOctaveBoundary('c', 'd', PitchDirection.DOWN));
        assert.equal(true,  crossesOctaveBoundary('d', 'c', PitchDirection.UP));
        assert.equal(false, crossesOctaveBoundary('d', 'c', PitchDirection.DOWN));
    });

});
